src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs

```diff
@@ -2580,8 +2580,13 @@ private void CompileCoalesceBinaryExpression(Expression expr)
                 // nullable value types with implicit (numeric) conversions which are allowed by Coalesce
                 // factory methods
 
-                Type nnLeftType = node.Left.Type.GetNonNullableType();
-                if (!TypeUtils.AreEquivalent(node.Type, nnLeftType))
+                Type typeToCompare = node.Left.Type;
+                if (!node.Type.IsNullableType())
+                {
+                    typeToCompare = node.Left.Type.GetNonNullableType();
+                }
```

Comment:
This part of the change is necessary because it would falsely decide the left hand needed to be converted and fixing that would let us know that if the left is converted we don't need to consider that the right might be. Skipping those unnecessary conversions otherwise is a free bonus.